### PR TITLE
[RFC] vim-patch:7.4.1102, 7.4.1110, 7.4.1832

### PR DESCRIPTION
**Pull request by jbradaric**
```
Problem:    Debugger has no stack backtrace support.
Solution:   Add "backtrace", "frame", "up" and "down" commands. (Alberto
            Fanjul, closes vim/vim#433)
```

https://github.com/vim/vim/commit/f1f60f859cdbb2638b3662ccf7b1d179865fe7dc

```
vim-patch:7.4.1110

Problem:    Test 108 fails when language is French.
Solution:   Force English messages. (Dominique Pelle)
```

https://github.com/vim/vim/commit/8c600052fabe4859470d9d0ba2ddd74a52ea9745

```
vim-patch:7.4.1832

Problem:    Memory leak in debug commands.
Solution:   Free memory before overwriting the pointer. (hint by Justin Keyes)
```

https://github.com/vim/vim/commit/dc303bce10c60a3314078ea168064552fadf01af

### CODE REVIEW

File: runtime/doc/repeat.txt [hunk 1]
```diff
@@ -467,16 +467,44 @@ Additionally, these commands can be used:
...
 - When you want to use the Ex command with the same name, prepend a colon:
   ":cont", ":next", ":finish" (or shorter).
 
+The backtrace shows the hierarchy of function calls, e.g.:
+	>bt ~
+	  3 function One[3] ~
+	  2 Two[3] ~
+	->1 Three[3] ~
+	  0 Four ~
+	line 1: let four = 4 ~
```
Comment (justinmk) [inline runtime/doc/repeat.txt]: It seems like `scripts/vim-patch.sh` or something in-between is turning tabs into single spaces.

File: src/nvim/ex_cmds2.c [hunk 2]
```diff
@@ -116,6 +116,11 @@ struct source_cookie {
 static int debug_greedy = FALSE;        /* batch mode debugging: don't save
                                            and restore typeahead. */
 
+static int get_maxbacktrace_level(void);
+static void do_setdebugtracelevel(char_u *arg);
+static void do_checkbacktracelevel(void);
+static void do_showbacktrace(char_u *cmd);
```
Comment (justinmk) [inline src/nvim/ex_cmds2.c]: These forward declarations aren't needed because of this line:

```
#ifdef INCLUDE_GENERATED_DECLARATIONS
# include "ex_cmds2.c.generated.h"
#endif
```

File: runtime/doc/repeat.txt (hunk 1 above)
Comment (jbradaric) [inline runtime/doc/repeat.txt]: @justinmk What do you mean? Each of these lines has a tab at the beginning.

File: src/nvim/ex_cmds2.c (hunk 2 above)
Comment (jbradaric) [inline src/nvim/ex_cmds2.c]: @justinmk Thanks, I've removed the declarations.

File: runtime/doc/repeat.txt (hunk 1 above)
Comment (justinmk) [inline runtime/doc/repeat.txt]: Hmm. I think GH has some display bugs.

**Comment by justinmk** [comment]
Build failures, somehow.

**Comment by jbradaric** [comment]
> Error: Error fetching file: Failed downloading https://luarocks.org/mpack-1.0.2-0.rockspec
make[2]: *** [usr/lib/luarocks/rocks/mpack] Error 1
make[1]: *** [CMakeFiles/mpack.dir/all] Error 2
make: *** [all] Error 2

**Comment by justinmk** [comment]
ASAN failure looks legitimate:

```
=================================================================
==25883==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 1500 byte(s) in 15 object(s) allocated from:
    #0 0x4f2282 in malloc (/home/travis/build/neovim/neovim/build/bin/nvim+0x4f2282)
    #1 0x2077d15 in try_malloc /home/travis/build/neovim/neovim/src/nvim/memory.c:59:15
    #2 0x2078761 in xmalloc /home/travis/build/neovim/neovim/src/nvim/memory.c:93:15
    #3 0x16f575f in alloc_cmdbuff /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:2074:20
    #4 0x16489f5 in command_line_enter /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:188:3
    #5 0x16439c3 in getcmdline /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:1635:10
    #6 0x1655500 in getcmdline_prompt /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:1663:7
    #7 0x12fd283 in do_debug /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:192:15
    #8 0x13e1154 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:909:7
    #9 0xc18e9b in call_user_func /home/travis/build/neovim/neovim/src/nvim/eval.c:20502:3
    #10 0xb758f7 in call_func /home/travis/build/neovim/neovim/src/nvim/eval.c:7239:11
    #11 0xbae0c6 in get_func_tv /home/travis/build/neovim/neovim/src/nvim/eval.c:7105:11
    #12 0xf593af in eval7 /home/travis/build/neovim/neovim/src/nvim/eval.c:4239:15
    #13 0xf4bf9b in eval6 /home/travis/build/neovim/neovim/src/nvim/eval.c:3960:7
    #14 0xf45ab0 in eval5 /home/travis/build/neovim/neovim/src/nvim/eval.c:3812:7
    #15 0xf36b92 in eval4 /home/travis/build/neovim/neovim/src/nvim/eval.c:3555:7
    #16 0xf346d0 in eval3 /home/travis/build/neovim/neovim/src/nvim/eval.c:3475:7
    #17 0xf32490 in eval2 /home/travis/build/neovim/neovim/src/nvim/eval.c:3412:7
    #18 0xb60a0d in eval1 /home/travis/build/neovim/neovim/src/nvim/eval.c:3345:7
    #19 0xb59b49 in eval0 /home/travis/build/neovim/neovim/src/nvim/eval.c:3307:9
    #20 0xb816a6 in ex_let /home/travis/build/neovim/neovim/src/nvim/eval.c:1407:9
    #21 0x14210c0 in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2191:15
    #22 0x13d3463 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
    #23 0xc18e9b in call_user_func /home/travis/build/neovim/neovim/src/nvim/eval.c:20502:3
    #24 0xb758f7 in call_func /home/travis/build/neovim/neovim/src/nvim/eval.c:7239:11
    #25 0xbae0c6 in get_func_tv /home/travis/build/neovim/neovim/src/nvim/eval.c:7105:11
    #26 0xf593af in eval7 /home/travis/build/neovim/neovim/src/nvim/eval.c:4239:15
    #27 0xf4bf9b in eval6 /home/travis/build/neovim/neovim/src/nvim/eval.c:3960:7
    #28 0xf45ab0 in eval5 /home/travis/build/neovim/neovim/src/nvim/eval.c:3812:7
    #29 0xf36b92 in eval4 /home/travis/build/neovim/neovim/src/nvim/eval.c:3555:7

Direct leak of 700 byte(s) in 7 object(s) allocated from:
    #0 0x4f2282 in malloc (/home/travis/build/neovim/neovim/build/bin/nvim+0x4f2282)
    #1 0x2077d15 in try_malloc /home/travis/build/neovim/neovim/src/nvim/memory.c:59:15
    #2 0x2078761 in xmalloc /home/travis/build/neovim/neovim/src/nvim/memory.c:93:15
    #3 0x16f575f in alloc_cmdbuff /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:2074:20
    #4 0x16489f5 in command_line_enter /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:188:3
    #5 0x16439c3 in getcmdline /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:1635:10
    #6 0x1655500 in getcmdline_prompt /home/travis/build/neovim/neovim/src/nvim/ex_getln.c:1663:7
    #7 0x12fd283 in do_debug /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:192:15
    #8 0x13054c7 in dbg_check_breakpoint /home/travis/build/neovim/neovim/src/nvim/ex_cmds2.c:482:7
    #9 0x13f446e in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:1459:3
    #10 0x13d3463 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
    #11 0xc18e9b in call_user_func /home/travis/build/neovim/neovim/src/nvim/eval.c:20502:3
    #12 0xb758f7 in call_func /home/travis/build/neovim/neovim/src/nvim/eval.c:7239:11
    #13 0xbae0c6 in get_func_tv /home/travis/build/neovim/neovim/src/nvim/eval.c:7105:11
    #14 0xf593af in eval7 /home/travis/build/neovim/neovim/src/nvim/eval.c:4239:15
    #15 0xf4bf9b in eval6 /home/travis/build/neovim/neovim/src/nvim/eval.c:3960:7
    #16 0xf45ab0 in eval5 /home/travis/build/neovim/neovim/src/nvim/eval.c:3812:7
    #17 0xf36b92 in eval4 /home/travis/build/neovim/neovim/src/nvim/eval.c:3555:7
    #18 0xf346d0 in eval3 /home/travis/build/neovim/neovim/src/nvim/eval.c:3475:7
    #19 0xf32490 in eval2 /home/travis/build/neovim/neovim/src/nvim/eval.c:3412:7
    #20 0xb60a0d in eval1 /home/travis/build/neovim/neovim/src/nvim/eval.c:3345:7
    #21 0xb59b49 in eval0 /home/travis/build/neovim/neovim/src/nvim/eval.c:3307:9
    #22 0xb816a6 in ex_let /home/travis/build/neovim/neovim/src/nvim/eval.c:1407:9
    #23 0x14210c0 in do_one_cmd /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:2191:15
    #24 0x13d3463 in do_cmdline /home/travis/build/neovim/neovim/src/nvim/ex_docmd.c:601:20
    #25 0xc18e9b in call_user_func /home/travis/build/neovim/neovim/src/nvim/eval.c:20502:3
    #26 0xb758f7 in call_func /home/travis/build/neovim/neovim/src/nvim/eval.c:7239:11
    #27 0xbae0c6 in get_func_tv /home/travis/build/neovim/neovim/src/nvim/eval.c:7105:11
    #28 0xf593af in eval7 /home/travis/build/neovim/neovim/src/nvim/eval.c:4239:15
    #29 0xf4bf9b in eval6 /home/travis/build/neovim/neovim/src/nvim/eval.c:3960:7

SUMMARY: AddressSanitizer: 2200 byte(s) leaked in 22 allocation(s).
```

File: src/nvim/ex_cmds2.c [hunk 3]
```diff
@@ -259,7 +292,28 @@ void do_debug(char_u *cmd)
...
+          }
+          continue;
+        case CMD_UP:
+          debug_backtrace_level++;
+          do_checkbacktracelevel();
+          continue;
+        case CMD_DOWN:
+          debug_backtrace_level--;
+          do_checkbacktracelevel();
+          continue;
```
Comment (justinmk) [inline src/nvim/ex_cmds2.c]: The introduction of `continue` here and above means `xfree(cmdline);` is never hit.

If you fix this memory leak please make it a separate commit. cc @chrisbra

**Comment by jbradaric** [comment]
@justinmk I've fixed the memory leak.

**Comment by justinmk** [comment]
QB failure seems invalid:

> Step 'master>buildall>build-node?testNode=macosx-64' is failed: Failed to delete file '/Users/quickbuild/tmp/buildagent/workspace/root/neovim/pull-requests-automated/.deps/usr/lib/luarocks/rocks/busted/2.0.rc11-0/bin/unit-test-directory/test.file'.

**Comment by justinmk** [comment]
@jbradaric Bram found a slightly cleaner fix (my name is on the commit though your name was on the patch I linked; next time, I will explicitly mention the author to avoid this): https://github.com/vim/vim/commit/dc303bce10c60a3314078ea168064552fadf01af

**Comment by jbradaric** [comment]
@justinmk No problem, as long as the bug is fixed :) I've also applied Bram's patch.

**Comment by jszakmeister** [comment]
> QB failure seems invalid:

> > Step 'master>buildall>build-node?testNode=macosx-64' is failed: Failed to delete file '/Users/quickbuild/tmp/buildagent/workspace/root/neovim/pull-requests-automated/.deps/usr/lib/luarocks/rocks/busted/2.0.rc11-0/bin/unit-test-directory/test.file'.

This was due to a test polluting the workspace and leaving files with awkward permissions, and QB not being able to handle it.  Specifically, a directory was left without execute permissions, so QB was not allowed to descend into it to clean the workspace.  I had to go back and remove the offending files from the nodes.  I was even tempted to add some other mechanism to help fix the issue, but I don't like covering up these kinds of problems either since they affect development on the desktop too.

**Comment by justinmk** [comment]
@jszakmeister What was the name of the directory? Then we might track down which test is causing the issue.

**Comment by jszakmeister** [comment]
Under `unit-test-directory` there was another directory named `test.file`.  It's tweaking the file permissions so that you can't descend into it (removing the executable bit).  I think it's in the fs_spec tests somewhere.
